main/src/cgeo/geocaching/Image.java

```diff
@@ -117,6 +364,68 @@ public boolean isLocalFile() {
      * @return the local file
      */
     public File localFile() {
-        return FileUtils.urlToFile(url);
+        return FileUtils.urlToFile(uri.toString());
+    }
+
+    /**
+     * Check if the image exists locally.
+     * Return False if Image is not local.
+     * Todo: May check if we have a cached Image for remote Uri
+     *
+     * @return
+     *          True if image exists on local filesystem
+     */
+    public boolean existsLocal() {
+        if (!isLocalFile()) {
+            return false;
+        }
+        return new File(getPath()).exists();
+    }
+
+    /**
+     * Check if the image Uri is Empty.
+     *
+     * @return
+     *          True if Uri is Empty or blank
+     */
+    public boolean isEmpty() {
+        return uri.equals(Uri.EMPTY) || StringUtils.isBlank(uri.toString());
+    }
+
+    /**
+     * Compare two Images.
+     *
+     * @param o
+     *          The Object to compare
+     * @return
+     *          True if all fields match
+     */
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final Image image = (Image) o;
+
+        if (!uri.equals(image.uri)) {
+            return false;
+        }
+        if (title != null ? !title.equals(image.title) : image.title != null) {
+            return false;
+        }
+        return !(description != null ? !description.equals(image.description) : image.description != null);
+
+    }
+
+    @Override
+    public int hashCode() {
+        int result = uri.hashCode();
```

Comment:
Have you looked at class `HashCodeBuilder` from the commons lang library (that we use)? It helps you building a hashcode, and its `append()` method supports `null` so it does not require any explicit testing. It should let you do something like (untested):

```java
   return new HashCodeBuilder().append(uri).append(title).append(description).build();
```